presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java

```diff
@@ -0,0 +1,258 @@
+            throw new IndexOutOfBoundsException("Invalid position " + position + " in block with " + positionCount + " positions");
+        }
+
+        int startValueOffset = getOffset(position);
+        int endValueOffset = getOffset(position + length);
+        Block newValues;
+        Block newOffsets;
+        Slice newValueIsNull;
+        int newOffsetBase;
+        if (compact) {
+            newValues = getValues().copyRegion(startValueOffset, endValueOffset - startValueOffset);
+            newOffsets = getOffsets().copyRegion(position, length);
+            newValueIsNull = Slices.copyOf(getValueIsNull(), position, length);
+            newOffsetBase = startValueOffset + getOffsetBase();
+            // In the case of compact, one could argue that is is beneficial to always set startValueOffset to 0, and adjust newOffsets accordingly.
+            // That would result in a unique in-memory representation for ArrayBlock when compacted.
+        }
+        else {
+            if (position == 0 && length == positionCount) {
+                // It is incorrect to pull up this `if` because child blocks may be compact-able when this if condition is satisfied
+                return this;
+            }
+            else {
+                newValues = getValues().getRegion(startValueOffset, endValueOffset - startValueOffset);
+                newOffsets = getOffsets().getRegion(position, length);
+                newValueIsNull = getValueIsNull().slice(position, length);
+                newOffsetBase = startValueOffset + getOffsetBase();
+            }
+        }
+        return new ArrayBlock(newValues, newOffsets, newOffsetBase, newValueIsNull);
+    }
+
+    @Override
+    public int getLength(int position)
+    {
+        return getOffset(position + 1) - getOffset(position);
+    }
+
+    @Override
+    public int getSizeInBytes()
+    {
+        return Integer.BYTES + getValues().getSizeInBytes() + getOffsets().getSizeInBytes() + getValueIsNull().length();
+    }
+
+    @Override
+    public int getRetainedSizeInBytes()
+    {
+        return Integer.BYTES + getValues().getRetainedSizeInBytes() + getOffsets().getRetainedSizeInBytes() + getValueIsNull().getRetainedSize();
+    }
+
+    @Override
+    public <T> T getObject(int position, Class<T> clazz)
+    {
+        if (clazz != Block.class) {
+            throw new IllegalArgumentException("clazz must be Block.class");
+        }
+        checkReadablePosition(position);
+
+        int positionCount = getPositionCount();
+        if (position < 0 || position >= positionCount) {
+            throw new IndexOutOfBoundsException("Invalid position " + position + " in block with " + positionCount + " positions");
+        }
+        int startValueOffset = getOffset(position);
+        int endValueOffset = getOffset(position + 1);
+        return (T) getValues().getRegion(startValueOffset, endValueOffset - startValueOffset);
```

Comment:
I have had this question about warnings for quite some time. How do you see the warnings?